Std/Tactic/LibrarySearch.lean

```diff
@@ -0,0 +1,531 @@
+| .internal id _ => id == abortSpeculationId
+| _ => false
+
+/--
+Sequentially invokes a tactic `act` on each value in candidates on the current state.
+
+The tactic `act` should return a list of meta-variables that still need to be resolved.
+If this list is empty, then no variables remain to be solved, and `tryOnEach` returns
+`none` with the environment set so each goal is resolved.
+
+If the action throws an internal exception with the `abortSpeculationId` id then
+further computation is stoped and intermediate results returned. If any other
+exception is thrown, then it is silently discarded.
+-/
+def tryOnEach
+    (act : Candidate → MetaM (List MVarId))
+    (candidates : Array Candidate) :
+    MetaM (Option (Array (List MVarId × MetavarContext))) := do
+  let mut a := #[]
+  let s ← saveState
+  for c in candidates do
+    match ← (tryCatch (Except.ok <$> act c) (pure ∘ Except.error)) with
+    | .error e =>
+      restoreState s
+      if isAbortSpeculation e then
+        break
+    | .ok remaining =>
+      if remaining.isEmpty then
+        return none
+      let ctx ← getMCtx
+      restoreState s
+      a := a.push (remaining, ctx)
+  return (.some a)
+
+/--
+Return an action that returns true when  the remaining heartbeats is less
+than the currently remaining heartbeats * leavePercent / 100.
+-/
+def mkHeartbeatCheck (leavePercent : Nat) : MetaM (MetaM Bool) := do
+  let maxHB ← getMaxHeartbeats
+  let hbThreshold := (← getRemainingHeartbeats) * leavePercent / 100
+  -- Return true if we should stop
+  pure $
+    if maxHB = 0 then
+      pure false
+    else do
+      return (← getRemainingHeartbeats) < hbThreshold
+
+/-- Shortcut for calling `solveByElim`. -/
+def solveByElim (required : List Expr) (exfalso : Bool) (goals : List MVarId) (maxDepth : Nat) := do
+  -- There is only a marginal decrease in performance for using the `symm` option for `solveByElim`.
+  -- (measured via `lake build && time lake env lean test/librarySearch.lean`).
+  let cfg : SolveByElim.Config :=
+    { maxDepth, exfalso := exfalso, symm := true, commitIndependentGoals := true }
+  let ⟨lemmas, ctx⟩ ← SolveByElim.mkAssumptionSet false false [] [] #[]
+  let cfg := if !required.isEmpty then cfg.requireUsingAll required else cfg
+  SolveByElim.solveByElim cfg lemmas ctx goals
+
+/-- State for resolving imports -/
+private def LibSearchState := IO.Ref (Option CandidateFinder)
+
+private initialize LibSearchState.default : IO.Ref (Option CandidateFinder) ← do
+  IO.mkRef .none
+
+private instance : Inhabited LibSearchState where
+  default := LibSearchState.default
+
+private initialize ext : EnvExtension LibSearchState ←
+  registerEnvExtension (IO.mkRef .none)
+
+private def librarySearchEmoji : Except ε (Option α) → String
+| .error _ => bombEmoji
+| .ok (some _) => crossEmoji
+| .ok none => checkEmoji
+
+/--
+Try to solve the goal either by:
+* calling `solveByElim`
+* or applying a library lemma then calling `solveByElim` on the resulting goals.
```

Comment:
```suggestion
Try to solve the goal either by:
* calling `tactic true`
* or applying a library lemma then calling `tactic false` on the resulting goals.

Typically here `tactic` is `solveByElim`,
with the `Bool` flag indicating whether it may retry with `exfalso`.
```